packages/core/src/codewhisperer/tracker/userWrittenCodeTracker.ts

```diff
@@ -0,0 +1,196 @@
+                                programmingLanguage: {
+                                    languageName: runtimeLanguageContext.toRuntimeLanguage(language),
+                                },
+                                acceptedCharacterCount: 0,
+                                totalCharacterCount: 0,
+                                timestamp: new Date(Date.now()),
+                                userWrittenCodeCharacterCount: charCount,
+                                userWrittenCodeLineCount: lineCount,
+                            },
+                        },
+                    })
+                    .then()
+                    .catch((error) => {
+                        let requestId: string | undefined
+                        if (isAwsError(error)) {
+                            requestId = error.requestId
+                        }
+                        getLogger().debug(
+                            `Failed to sendTelemetryEvent, requestId: ${requestId ?? ''}, message: ${error.message}`
+                        )
+                    })
+            }
+        }
+    }
+
+    private tryStartTimer() {
+        if (this._timer !== undefined) {
+            return
+        }
+        if (!this.isActive()) {
+            getLogger().debug(`Skip emiting code contribution metric. Telemetry disabled or not logged in. `)
+            this.reset()
+            return
+        }
+        const startTime = performance.now()
+        this._timer = setTimeout(() => {
+            try {
+                const currentTime = performance.now()
+                const delay: number = UserWrittenCodeTracker.defaultCheckPeriodMillis
+                const diffTime: number = startTime + delay
+                if (diffTime <= currentTime) {
+                    if (this._qUsageCount <= 0) {
+                        getLogger().debug(`Skip emiting code contribution metric. There is no active Amazon Q usage. `)
+                        return
+                    }
+                    if (this._userWrittenNewCodeCharacterCount.size === 0) {
+                        getLogger().debug(`Skip emiting code contribution metric. There is no new code added. `)
+                        return
+                    }
+                    this.emitCodeContributions()
+                }
+            } catch (e) {
+                getLogger().verbose(`Exception Thrown from QCodeGenTracker: ${e}`)
+            } finally {
+                this.reset()
+            }
+        }, UserWrittenCodeTracker.defaultCheckPeriodMillis)
+    }
+
+    private countNewLines(str: string) {
+        return str.split('\n').length - 1
+    }
+
+    public onTextDocumentChange(e: vscode.TextDocumentChangeEvent) {
+        // do not count code written by Q as user written code
+        if (
+            !runtimeLanguageContext.isLanguageSupported(e.document.languageId) ||
```

Comment:
double checking `runtimeLanguageContext.isLanguageSupported()` here refers to language supported by `inline completion`, and I thought languages supported by Chat is a super set of this?

Reply:
yes